apps/session-manager/src/controllers/fast-login.ts

```diff
@@ -1,16 +1,125 @@
 import { GenerateNonceResponse } from "../generated/fast-login-api/GenerateNonceResponse";
 import { assertNever, readableProblem } from "../utils/errors";
 import { getFnFastLoginAPIClient } from "../repositories/fast-login-api";
+import {
+  BPDToken,
+  FIMSToken,
+  MyPortalToken,
+  SessionToken,
+  UserTokens,
+  WalletToken,
+  ZendeskToken,
+} from "../types/token";
+import { getNewTokenAsync } from "../services/token";
+import { User } from "../types/user";
+import { ResLocals } from "../utils/express";
+import { withLollipopLocals } from "../utils/lollipop";
+import { FastLoginResponse as LCFastLoginResponse } from "../generated/fast-login-api/FastLoginResponse";
+import { makeProxyUserFromSAMLResponse } from "../utils/spid";
+import { isBlockedUser, set } from "../services/redis-session-storage";
+import { FastLoginResponse } from "../types/fast-login";
+import { RedisClientSelectorType } from "../types/redis";
+import { RedisRepositoryDeps } from "../repositories/redis";
+import { WithIP } from "../utils/network";
+import { SESSION_ID_LENGTH_BYTES, SESSION_TOKEN_LENGTH_BYTES } from "./session";
+
+const generateSessionTokens = (
+  userFiscalCode: FiscalCode,
+  redisClientSelector: RedisClientSelectorType,
+): TE.TaskEither<
+  IResponseErrorInternal | IResponseErrorUnauthorized,
+  UserTokens
+> => {
+  // note: since we have a bunch of async operations that don't depend on
+  //       each other, we can run them in parallel
+  const tokenTasks = {
+    // authentication token for BPD
+    bpd_token: () =>
+      getNewTokenAsync(SESSION_TOKEN_LENGTH_BYTES) as Promise<BPDToken>,
+    // authentication token for FIMS
+    fims_token: () =>
+      getNewTokenAsync(SESSION_TOKEN_LENGTH_BYTES) as Promise<FIMSToken>,
+    // authentication token for MyPortal
+    myportal_token: () =>
+      getNewTokenAsync(SESSION_TOKEN_LENGTH_BYTES) as Promise<MyPortalToken>,
+    // authentication token for app backend
+    session_token: () =>
+      getNewTokenAsync(SESSION_TOKEN_LENGTH_BYTES) as Promise<SessionToken>,
+    // unique ID for tracking the user session
+    session_tracking_id: () => getNewTokenAsync(SESSION_ID_LENGTH_BYTES),
+    // authentication token for pagoPA
+    wallet_token: () =>
+      getNewTokenAsync(SESSION_TOKEN_LENGTH_BYTES) as Promise<WalletToken>,
+    // authentication token for Zendesk
+    zendesk_token: () =>
+      getNewTokenAsync(SESSION_TOKEN_LENGTH_BYTES) as Promise<ZendeskToken>,
+  };
+
+  return pipe(
+    // ask the session storage whether this user is blocked
+    isBlockedUser({
+      selector: redisClientSelector,
+      fiscalCode: userFiscalCode,
+    }),
+    TE.mapLeft(() => ResponseErrorInternal(`Error while validating user`)),
+    TE.chainW((isUserBlocked) =>
+      isUserBlocked
+        ? TE.left(ResponseErrorUnauthorized("User is blocked"))
+        : TE.right(true),
+    ),
+    TE.chainW(() => pipe(tokenTasks, AP.sequenceS(T.ApplyPar), TE.fromTask)),
```

Comment:
ditto

Reply:
see comment above

Reply:
Add a TODO that remember us to refactor this part removing the async token generation

Reply:
done in https://github.com/pagopa/io-auth-n-identity-domain/pull/22/commits/efaf8f33c282bf83e278da21b6f4c681efe06acd